onnx/onnx_backend.h

```diff
@@ -0,0 +1,801 @@
+ *                                      couldn't parse the serialized protobuf
+ *                                      as an ONNX ModelProto message.
+ * @retval ONNX_STATUS_INVALID_MODEL The function call failed because the parsed
+ *                                   ModelProto message does not satisfy ONNX
+ *                                   requirements and constraints.
+ * @retval ONNX_STATUS_UNSUPPORTED_VERSION The function call failed because the
+ *                                         ONNX IR version or operator version
+ *                                         is not supported by the backend.
+ * @retval ONNX_STATUS_UNSUPPORTED_OPERATOR The function call failed because one
+ *                                          of the operators in the model graph
+ *                                          is not supported by the backend.
+ * @retval ONNX_STATUS_UNSUPPORTED_PARAMETER The function call failed because
+ *                                           the backend does not support the
+ *                                           particular parameter values in one
+ *                                           of the operators.
+ * @retval ONNX_STATUS_NO_SYSTEM_MEMORY The function call failed because the
+ *                                      backend could not allocate enough
+ *                                      system memory to parse, analyze, and
+ *                                      initialize the model graph.
+ * @retval ONNX_STATUS_NO_SYSTEM_RESOURCES The function call failed due to
+ *                                         insufficient non-memory system
+ *                                         resources (e.g. file handles) to
+ *                                         initialize the graph.
+ * @retval ONNX_STATUS_NO_DEVICE_MEMORY The function call failed due to
+ *                                      insufficient backend-specific memory to
+ *                                      initialize the graph.
+ * @retval ONNX_STATUS_NO_DEVICE_RESOURCES The function call failed due to
+ *                                         insufficient non-memory
+ *                                         backend-specific resources (e.g.
+ *                                         command queues) to initialize the
+ *                                         graph.
+ */
+ONNX_PUBLIC onnxStatus ONNX_ABI ONNX_SYMBOL_NAME(onnxInitGraph)(
+    onnxBackend backend,
+    size_t onnxModelSize,
+    const void* onnxModel,
+    uint32_t weightsCount,
+    const onnxTensorDescriptor* weightDescriptors,
+    onnxGraph* graph);
+
+/**
+ * Set locations for inputs and outputs of an ONNX graph.
+ *
+ * The caller MUST ensure that the memory buffers specified for input and output
+ * tensors remain accessible for the life-time of the ONNX graph.
+ *
+ * @param graph - graph handle created by onnxInitGraph.
+ * @param inputsCount - number of elements in the inputDescriptors array.
```

Comment:
What happens if `inputsCount` or `outputCount` don't match the number of required inputs/outputs?

Reply:
Undefined behavior

Reply:
So actually there is no reason to pass these in at all because the values are predetermined anyhow by the model? Hm. I'm unsure, it probably still makes sense to keep it as-is (i.e. keep the arguments there) to allow backends to do sanity checking.

Reply:
Now I get that you asked a different question. If `inputsCounts`/`outputsCount` doesn't match the number of GraphProto.inputs (excluding static weights) and GraphProto.outputs, the function fails with `ONNX_STATUS_UNIDENTIFIED_NAME`

Reply:
If `ModelProto.graph.input < inputDescriptors`, then either some input descriptors are duplicate, or they contain a name that is not in `ModelProto.graph.input`. Either way `onnxSetGraphIO` would fail with `ONNX_STATUS_INVALID_NAME`.

Reply:
If `ModelProto.graph.input < inputDescriptors`, then either some input descriptors are duplicate, or they contain a name that is not in `ModelProto.graph.input`. Either way `onnxSetGraphIO` would fail with `ONNX_STATUS_INVALID_NAME`.